README.md
# airflow-looker

This is a collection of [Airflow](https://airflow.apache.org/) extensions to provide integration with [Looker](https://www.looker.com).

```py
with DAG(dag_id='looker_update_datagroup', schedule_interval='@daily') as dag:
  LookerUpdateDataGroupByIDOperator(
    datagroup_id=123
  )
```

## Installation

Install from PyPI:

```sh
pip install airflow-looker
```

## Usage

There is one operator currently implemented:

* `LookerUpdateDataGroupByIDOperator`
  * Calls the [`update_datagroup` API](https://docs.looker.com/reference/api-and-integration/api-reference/v3.0/datagroup#update_datagroup). Accepts the following arguments:
    * `datagroup_id`
      * The ID of the datagroup to update. Required.
    * `stale_before`
      * Timestamp before which cache entries are considered stale. Defaults to now.

You can also use the hook directly. The two methods that are implemented for use are:

* `call`
  * Call the Looker API. Accepts the following arguments:
    * `method`
      * The method of the call (`GET`, `POST`, etc). Required.
    * `endpoint`
      * The endpoint to be called i.e. `looks/run/1`. Required.
    * `data`
      * Payload to be uploaded or request parameters. Required.
    * `headers`
      * Additional headers to be passed through as a dictionary. Optional.
* `get_look_sql`
  * Gets an SQL query from a Looker look resource and returns the SQL as a string. Accepts the following arguments:
    * `look_id`
      * Unique identifier for a look resource. Required.

### Connection

To use either the operator or the hook you need to pass in a connection ID. This connection needs to have the the host, the login (`client_id`) and the password (`client_secret`) defined.

To create a connection, follow the [Airflow documentation](https://airflow.apache.org/docs/stable/howto/connection/index.html).

## Building Locally

To install from the repository, it's recommended to first create a virtual environment:

```bash
python3 -m venv .venv

source .venv/bin/activate
```

Install using `pip`:

```bash
pip install .
```

## Testing

To run tests locally, first create a virtual environment (see [Building Locally](https://github.com/gocardless/airflow-looker#building-locally) section)

Install dependencies:

```bash
pip install -e .[dev]
```

Run the tests:

```bash
python -m pytest tests/ -sv
```

## Code style

This project uses [flake8](https://flake8.pycqa.org/en/latest/).

To check your code, first create a virtual environment (see [Building Locally](https://github.com/gocardless/airflow-looker#building-locally) section):

```bash
python -m flake8 airflow_looker/ tests/ setup.py
```

## License & Contributing

* This is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).
* Bug reports and pull requests are welcome on GitHub at https://github.com/gocardless/airflow-looker.

GoCardless ♥ open source. If you do too, come [join us](https://gocardless.com/about/jobs).


CHANGELOG.md
# v0.0.1

Initial release with a single operator:

* `LookerUpdateDataGroupByIDOperator`
  * Calls the [`update_datagroup` API](https://docs.looker.com/reference/api-and-integration/api-reference/v3.0/datagroup#update_datagroup).

And the following hook methods:

* `call`
  * Call the Looker API
* `get_look_sql`
  * Gets a SQL query from a Looker look resource and returns the SQL as a string.


tests/hooks/test_looker_hook.py
import io
import json
import requests
import unittest
import requests_mock
from unittest import mock
from requests import Response
from airflow import AirflowException
from airflow.hooks.base_hook import BaseHook
from airflow.models import Connection
from airflow_looker.hooks.looker_hook import LookerHook


class TestLookerHook(unittest.TestCase):
    def setUp(self):
        self.looker_host = "http://127.0.0.1:8080/"
        self.airflow_looker_conn_id = "connexion_defacto"
        self.looker_airflow_connection = Connection(
            conn_id=self.airflow_looker_conn_id,
            conn_type='http',
            host=self.looker_host,
            login='looker_api_client_id',
            password='looker_api_client_secret'
        )
        self.looker_auth_token = "fancy-pancy-access-token"
        self.login_response_payload = {"access_token": self.looker_auth_token}
        self.default_hook = LookerHook()
        session = requests.Session()
        adapter = requests_mock.Adapter()
        session.mount('mock', adapter)

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_create_custom_connection(self, mock_request, mock_get_connection):
        looker_auth_url = "{}{}".format(self.looker_host, "login")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )
        mock_get_connection.return_value = self.looker_airflow_connection
        hook = LookerHook(looker_conn_id=self.airflow_looker_conn_id)
        conn = hook.get_conn()
        self.assertEqual(self.airflow_looker_conn_id, hook.looker_conn_id)
        self.assertEqual(self.looker_host, hook.api_endpoint)
        self.assertEqual("token {}".format(self.looker_auth_token), conn.headers['Authorization'])

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_create_default_connection(self, mock_request, mock_get_connection):
        looker_auth_url = "{}{}".format(self.looker_host, "login")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        self.looker_airflow_connection.conn_id = "looker_default"
        mock_get_connection.return_value = self.looker_airflow_connection
        conn = self.default_hook.get_conn()
        self.assertEqual("looker_default", self.default_hook.looker_conn_id)
        self.assertEqual(self.looker_host, self.default_hook.api_endpoint)
        self.assertEqual("token {}".format(self.looker_auth_token), conn.headers['Authorization'])

    @mock.patch.object(BaseHook, "get_connection")
    def test_create_hook_without_airflow_connection(self, mock_get_connection):
        # Here a default connection will be created with looker_conn_id='looker_default' but wont have password, login
        # attributes
        mock_get_connection.return_value = Connection(
            conn_id=self.airflow_looker_conn_id,
            conn_type='http'
        )
        with self.assertRaises(AirflowException):
            self.default_hook.get_conn()

    @mock.patch.object(BaseHook, "get_connection")
    def test_create_hook_without_airflow_connection_password(self, mock_get_connection):
        mock_get_connection.return_value = Connection(
            conn_id=self.airflow_looker_conn_id,
            conn_type='http',
            host=self.looker_host,
            login='looker_api_client_id'
        )
        with self.assertRaises(AirflowException):
            self.default_hook.get_conn()

    @mock.patch.object(BaseHook, "get_connection")
    def test_create_hook_without_airflow_connection_login(self, mock_get_connection):
        mock_get_connection.return_value = Connection(
            conn_id=self.airflow_looker_conn_id,
            conn_type='http',
            host=self.looker_host,
            password='looker_api_client_secret'
        )
        with self.assertRaises(AirflowException):
            self.default_hook.get_conn()

    @mock.patch.object(BaseHook, "get_connection")
    def test_create_hook_without_airflow_connection_host(self, mock_get_connection):
        mock_get_connection.return_value = Connection(
            conn_id=self.airflow_looker_conn_id,
            conn_type='http',
            login='looker_api_client_id',
            password='looker_api_client_secret'
        )
        with self.assertRaises(AirflowException):
            self.default_hook.get_conn()

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_call_method_with_http_get(self, mock_request, mock_get_connection):
        mock_get_connection.return_value = self.looker_airflow_connection
        looker_auth_url = "{}{}".format(self.looker_host, "login")
        looks_url = "{}{}".format(self.looker_host, "looks")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        mock_request.get(
            looks_url,
            status_code=200,
            text='{"object":"looker_looks_resource"}',
            reason='OK'
        )
        response = self.default_hook.call(method="GET", endpoint="looks", data=None)
        self.assertEqual("GET", response.request.method)
        self.assertEqual(looks_url, response.request.url)
        self.assertEqual(200, response.status_code)

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_call_method_with_http_get_params(self, mock_request, mock_get_connection):
        mock_get_connection.return_value = self.looker_airflow_connection
        looker_auth_url = "{}{}".format(self.looker_host, "login")
        looks_url = "{}{}".format(self.looker_host, "looks")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        mock_request.get(
            looks_url,
            status_code=200,
            text='{"object":"looker_looks_resource"}',
            reason='OK'
        )
        param_key = "include"
        param_value = "obsolete"
        response = self.default_hook.call(method="GET", endpoint="looks", data={param_key: param_value})
        self.assertEqual("GET", response.request.method)
        self.assertEqual("{}?{}={}".format(looks_url, param_key, param_value), response.request.url)

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_call_method_with_http_head(self, mock_request, mock_get_connection):
        mock_get_connection.return_value = self.looker_airflow_connection
        looker_auth_url = "{}{}".format(self.looker_host, "login")
        looks_url = "{}{}".format(self.looker_host, "looks")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        mock_request.head(
            looks_url,
            status_code=200,
            text='{"object":"looker_looks_resource"}',
            reason='OK'
        )
        response = self.default_hook.call(method="HEAD", endpoint="looks", data=None)
        self.assertEqual("HEAD", response.request.method)
        self.assertEqual(looks_url, response.request.url)
        self.assertEqual(200, response.status_code)

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_call_method_with_request_headers(self, mock_request, mock_get_connection):
        mock_get_connection.return_value = self.looker_airflow_connection
        looker_auth_url = "{}{}".format(self.looker_host, "login")
        looks_url = "{}{}".format(self.looker_host, "looks")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        mock_request.head(
            looks_url,
            status_code=200,
            text='{"object":"looker_looks_resource"}',
            reason='OK'
        )
        request_headers = {"looker-api-version": "2020-01-01"}
        response = self.default_hook.call(method="HEAD", endpoint="looks", data=None, headers=request_headers)
        self.assertEqual(request_headers['looker-api-version'], response.request.headers['looker-api-version'])
        self.assertEqual("HEAD", response.request.method)
        self.assertEqual(looks_url, response.request.url)
        self.assertEqual(200, response.status_code)

    @requests_mock.mock()
    @mock.patch.object(BaseHook, "get_connection")
    def test_call_method_with_http_post(self, mock_request, mock_get_connection):
        mock_get_connection.return_value = self.looker_airflow_connection
        looker_auth_url = "{}{}".format(self.looker_host, "login")
        looks_url = "{}{}".format(self.looker_host, "looks")

        mock_request.post(
            looker_auth_url,
            status_code=200,
            text=json.dumps(self.login_response_payload),
            reason='OK'
        )

        mock_request.post(
            looks_url,
            status_code=200,
            text='{"object":"looker_looks_resource"}',
            reason='OK'
        )
        request_payload = '{"post_object":"looker_looks_resource"}'
        response = self.default_hook.call(method="POST", endpoint="looks", data=request_payload)
        self.assertEqual("POST", response.request.method)
        self.assertEqual(looks_url, response.request.url)
        self.assertEqual(json.dumps(request_payload), response.request.body)
        self.assertEqual(200, response.status_code)

    @mock.patch.object(LookerHook, 'call')
    def test_get_look_sql_correct_args(self, mock_request):
        mock_response = Response()
        mock_response.status_code = 200
        mock_sql_query = b'''
SELECT * FROM music_albums
WHERE artist = "cardi b"
'''
        mock_response.raw = io.BytesIO(mock_sql_query)
        mock_request.return_value = mock_response

        look_id = 1234
        query_fetch_method = 'GET'
        endpoint = "{}/{}/run/{}".format('api/3.0/looks', look_id, 'sql')

        query = self.default_hook.get_look_sql(look_id)
        mock_request.assert_called_once_with(
            data=None,
            endpoint=endpoint,
            method=query_fetch_method
        )
        self.assertEqual(mock_sql_query.decode('utf8'), query)

    @mock.patch.object(LookerHook, 'call')
    def test_get_look_sql_with_looker_unsuccessful_response(self, mock_request):
        mock_response = Response()
        mock_response.status_code = 400
        mock_sql_query = b'''
        SELECT * FROM music_albums
        WHERE artist = "cardi b"
        '''
        mock_response.raw = io.BytesIO(mock_sql_query)
        mock_request.return_value = mock_response

        with self.assertRaises(AirflowException):
            self.default_hook.get_look_sql(134)

    def test_get_look_sql_missing_look_id(self):
        with self.assertRaises(AirflowException):
            self.default_hook.get_look_sql()


suite = unittest.TestLoader().loadTestsFromTestCase(TestLookerHook)
unittest.TextTestRunner(verbosity=2).run(suite)


airflow_looker/operators/looker_operator.py
import time
from airflow_looker.hooks.looker_hook import LookerHook
from airflow.models import BaseOperator
from airflow.utils.decorators import apply_defaults


class LookerOperator(BaseOperator):
    ui_color = '#615286'

    @apply_defaults
    def __init__(self, looker_conn_id='looker_default', *args, **kwargs):
        super(LookerOperator, self).__init__(*args, **kwargs)
        self.looker_conn_id = looker_conn_id

    def _get_hook(self):
        return LookerHook(looker_conn_id=self.looker_conn_id)


class LookerUpdateDataGroupByIDOperator(LookerOperator):
    """
    Update a datagroup using the specified params.

    :param datagroup_id: The Datagroup ID to update. Required.
    :type datagroup_id: int
    :param looker_conn_id: reference to a specific Looker connection.
    :type looker_conn_id: string
    :param stale_before: The datagroup is stale if refreshed before this time. Defaults to now.
    :type stale_before: time
    """
    @apply_defaults
    def __init__(self, looker_conn_id='looker_default', datagroup_id=None, stale_before=time.time(), *args, **kwargs):
        super(LookerUpdateDataGroupByIDOperator, self).__init__(looker_conn_id=looker_conn_id, *args, **kwargs)
        self.datagroup_id = datagroup_id
        self.stale_before = stale_before

    def execute(self, context):
        # get hook and call
        looker = self._get_hook()

        endpoint = '{}/{}'.format('api/3.0/datagroups', self.datagroup_id)
        body = {
            'stale_before': self.stale_before
        }

        self.log.info("Calling: %s with body: %s", endpoint, body)
        looker.call(method='PATCH', endpoint=endpoint, data=body)


airflow_looker/hooks/looker_hook.py
import requests
import json
from urllib.parse import urljoin

from airflow.hooks.base_hook import BaseHook
from airflow.exceptions import AirflowException


class LookerHook(BaseHook):
    """
    A hook to talk to the Looker API over HTTP
    """
    def __init__(self,
                 looker_conn_id='looker_default',
                 verify=True):
        """
        :param looker_conn_id: connection that has the host i.e
        https://looker.company.com:19999/api/3.0/, the login (client_id) and
        the password (client_secret).
        :type looker_conn_id: string
        :param verify: true if we should verify API calls. Defaults to true.
        :type verify: boolean
        """
        self.looker_conn_id = looker_conn_id
        self.verify = verify
        self.api_endpoint = None

    def get_conn(self):
        """
        Returns http session for use with requests
        """
        conn = self.get_connection(self.looker_conn_id)

        if conn.host is None:
            _message = "Failed to initialize looker airflow connector, connection host not provided"
            self.logger.error(_message)
            raise AirflowException(_message)

        if conn.login is None:
            _message = "Failed to initialize looker Airflow connector, connection login not provided"
            self.logger.error(_message)
            raise AirflowException(_message)

        if conn.password is None:
            _message = "Failed to initialize looker Airflow connector, connection password not provided"
            self.logger.error(_message)
            raise AirflowException(_message)

        session = requests.Session()

        self.api_endpoint = conn.host
        token_request = requests.post(
            url=urljoin(self.api_endpoint, "login"),
            data={
                "client_id": conn.login,
                "client_secret": conn.password
            },
            verify=self.verify,
        )

        token = token_request.json()["access_token"]
        headers = {"Authorization": "token " + token}
        session.headers.update(headers)

        return session

    def call(self, method, endpoint, data, headers=None):
        """
        Call the Looker API and return results
        :param method: the method of the call (`GET`, `POST`, etc)
        :type method: str
        :param endpoint: the endpoint to be called i.e. looks/run/1
        :type endpoint: str
        :param data: payload to be uploaded or request parameters
        :type data: dict
        :param headers: additional headers to be passed through as a dictionary
        :type headers: dict
        """

        session = self.get_conn()
        url = urljoin(self.api_endpoint, endpoint)

        req = None
        if method == 'GET':
            # GET uses params
            req = requests.Request(method,
                                   url,
                                   params=data,
                                   headers=headers)
        elif method == 'HEAD':
            # HEAD doesn't use params
            req = requests.Request(method,
                                   url,
                                   headers=headers)
        else:
            # Others use data
            req = requests.Request(method,
                                   url,
                                   data=json.dumps(data),
                                   headers=headers)

        prepped_request = session.prepare_request(req)
        self.log.info("Sending '%s' to url: %s: %s", method, url, data)
        response = session.send(prepped_request, verify=self.verify)
        try:
            response.raise_for_status()
        except requests.exceptions.HTTPError:
            self.log.error("HTTP error: %s", response.reason)
            self.log.error(response.text)
            raise AirflowException(str(response.status_code) + ":" + response.reason)
        return response

    def get_look_sql(self, look_id=None):
        """
        Gets a SQL query from a Looker look resource
        :param look_id: unique identifier for a look resource
        :type look_id: int
        :return: sql string
        """
        if look_id is None:
            _message = "No look_id provided"
            self.log.error(_message)
            raise AirflowException(_message)

        endpoint = '{}/{}/run/{}'.format('api/3.0/looks', look_id, 'sql')
        self.log.info("Fetching looker %s query", endpoint)
        response = self.call(method='GET', endpoint=endpoint, data=None)

        if response.status_code > 200:
            _message = 'Failed to fetch query {}'.format(response.reason)
            self.log.error(_message)
            raise AirflowException(_message)
        return response.text
